packages/@aws-cdk/aws-cognito-identitypool/lib/identitypool-user-pool-authentication-provider.ts

```diff
@@ -0,0 +1,85 @@
+import { randomBytes } from 'crypto';
+import {
+  IUserPoolIdentityProvider,
+  UserPoolClient,
+  UserPoolClientProps,
+} from '@aws-cdk/aws-cognito';
+import { DependableTrait } from '@aws-cdk/core';
+
+/**
+ * Represents a UserPoolAuthenticationProvider
+ */
+export interface IUserPoolAuthenticationProvider {
+  /**
+   * Client Id of the Associated User Pool Client
+   */
+  readonly clientId: string
+
+  /**
+   * Whether to disable the pool's default server side token check
+   */
+  readonly disableServerSideTokenCheck: boolean;
+
+  /**
+   * The identity providers associated with the UserPool
+   */
+  readonly identityProviders: IUserPoolIdentityProvider[];
+}
+
+/**
+ * Props for the User Pool Authentication Provider
+ */
+export interface UserPoolAuthenticationProviderProps extends UserPoolClientProps {
+
+  /**
+   * Setting this to true turns off identity pool checks for this user pool to make sure the user has not been globally signed out or deleted before the identity pool provides an OIDC token or AWS credentials for the user
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-identitypool-cognitoidentityprovider.html
+   * @default false
+   */
+  readonly disableServerSideTokenCheck?: boolean;
+}
+
+/**
+ * Defines a User Pool Authentication Provider
+ */
+export class UserPoolAuthenticationProvider implements IUserPoolAuthenticationProvider {
+
+  /**
+   * Client Id of the Associated User Pool Client
+   */
+  public readonly clientId: string
+
+  /**
+   * Whether to disable the pool's default server side token check
+   */
+  public readonly disableServerSideTokenCheck: boolean;
+
+  /**
+   * The identity providers associated with the UserPool
+   */
+  public readonly identityProviders: IUserPoolIdentityProvider[] = [];
+
+  constructor(props: UserPoolAuthenticationProviderProps) {
+    const client = this.configureUserPoolClient(props);
+    this.clientId = client.userPoolClientId;
+    this.identityProviders = props.userPool.identityProviders;
```

Comment:
@smguggen I'm actually not sure if this is correct. I _think_ that the IdentityPool.cognitoIdentityProviders requires the `providerName` to always be the UserPool provider name (i.e. `cognito-idp.us-east-1.amazonaws.com/us-east-1_123456789`)

So you would get a single `cognitoIdentityProvider` per UserPool/Client.

Reply:
@corymhall The docs do seem to suggest that the `userPoolProviderName` should be used, but if that were the case then there would be no reason for `IdentityPool.cognitoIdentityProviders` to be an array. So looping through the array of a user pool's `identityProviders` and using each `providerName` worked when I tested it. Let me look into that a little bit further.

Reply:
@corymhall Ok well I looked into it and it looks like we made need to add an issue because `UserPool.userPoolProviderName` refers to an Attribute that doesn't exist, at least not anymore. It references the L1 Construct's `attrProviderName` which tries to get the `ProviderName` attribute from the Cloud Formation User Pool, only as far as I can tell there is no `ProviderName` attribute for the user pool. There's a `ProviderName` for a `UserPoolIdentityProvider` but those aren't attached to a `UserPool` by default and they're attached to the list of `identityProviders` in the user pool anyway, so there's no need for a `ProviderName` for a user pool because there could be any number of providers (including zero) attached to it.

I'm not sure if in the past this was needed but now that user pools have a list of `identityProviders` attached to them, there seems to be no need for it anymore, and as I said before, setting `cognitoIdentityProviders` to an array including all of the `providerNames` of all of a user pool's identity providers works fine, so I think this is a case where the documentation needs to be updated and the property can be removed.

Reply:
And this needs to be investigated, but the above most likely applies to `UserPoolProviderUrl` too

Reply:
@smguggen In my testing I was only able to get it to work with the `UserPool.userPoolProviderName`. Can you add an integration test that uses providers, I think that will ultimately validate what the correct configuration should be.

Reply:
@corymhall The integration test already does that. The user pool inside of the L2 `IdentityPool.authenticationProviders` is added under the hood to the L1 `CfnIdentityPool.cognitoIdentityProviders`. It uses the `UserPoolAuthenticationProvider` class which provides the name from the identity provider rather than `UserPool.userPoolProviderName`. But `UserPool.userPoolProviderName` gets the name from the identity provider anyway, the problem is that on the CloudFormation template level there's a good chance it could cause a stack failure because it refers to a property that may not exist.